File: src/lib/api/export.ts
import { invoke } from '@tauri-apps/api/core'
import type { ExportRequest, ExportResult, DeviceExport, ExportCheckpoint } from '$lib/types'

/**
 * Export playlists to a USB device
 */
export async function exportToDevice(request: ExportRequest): Promise<ExportResult> {
	return invoke('export_playlists', { request })
}

/**
 * Get all exports for a device
 */
export async function getDeviceExports(deviceId: string): Promise<DeviceExport[]> {
	return invoke('get_device_exports', { deviceId })
}

/**
 * Cancel the current export operation
 */
export async function cancelExport(): Promise<void> {
	return invoke('cancel_export')
}

/**
 * Clean up a failed export by removing copied files
 */
export async function cleanupFailedExport(deviceId: string, mountPoint: string): Promise<void> {
	return invoke('cleanup_failed_export', { deviceId, mountPoint })
}

/**
 * Get a pending checkpoint for a device
 */
export async function getPendingCheckpoint(deviceId: string): Promise<ExportCheckpoint | null> {
	return invoke('get_pending_checkpoint', { deviceId })
}

/**
 * Delete a checkpoint
 */
export async function deleteCheckpoint(checkpointId: string): Promise<void> {
	return invoke('delete_checkpoint', { checkpointId })
}

/**
 * Resume a previously interrupted export
 */
export async function resumeExport(deviceId: string, mountPoint: string): Promise<ExportResult> {
	return invoke('resume_export', { deviceId, mountPoint })
}

File: src-tauri/src/commands/export.rs
use std::sync::Arc;

use tauri::{AppHandle, State};

use crate::error::Result;
use crate::models::export::ExportCheckpoint;
use crate::models::{DeviceExport, ExportRequest, ExportResult};
use crate::services::export::CheckpointService;
use crate::services::ExportService;

/// Export playlists to a USB device
#[tauri::command]
pub async fn export_playlists(
    request: ExportRequest,
    export_service: State<'_, Arc<ExportService>>,
    app_handle: AppHandle,
) -> Result<ExportResult> {
    export_service.export_playlists(&app_handle, request)
}

/// Get all exports for a device
#[tauri::command]
pub async fn get_device_exports(
    device_id: String,
    export_service: State<'_, Arc<ExportService>>,
) -> Result<Vec<DeviceExport>> {
    export_service.get_device_exports(&device_id)
}

/// Cancel the current export operation
#[tauri::command]
pub async fn cancel_export(export_service: State<'_, Arc<ExportService>>) -> Result<()> {
    export_service.cancel_export();
    Ok(())
}

/// Clean up a failed export by removing copied files
#[tauri::command]
pub async fn cleanup_failed_export(
    device_id: String,
    mount_point: String,
    export_service: State<'_, Arc<ExportService>>,
) -> Result<()> {
    export_service.cleanup_failed_export(&device_id, &mount_point)
}

/// Get a pending checkpoint for a device
#[tauri::command]
pub async fn get_pending_checkpoint(
    device_id: String,
    checkpoint_service: State<'_, Arc<CheckpointService>>,
) -> Result<Option<ExportCheckpoint>> {
    checkpoint_service.get_pending_checkpoint(&device_id)
}

/// Delete a checkpoint
#[tauri::command]
pub async fn delete_checkpoint(
    checkpoint_id: String,
    checkpoint_service: State<'_, Arc<CheckpointService>>,
) -> Result<()> {
    checkpoint_service.delete_checkpoint(&checkpoint_id)
}

/// Resume a previously interrupted export
#[tauri::command]
pub async fn resume_export(
    device_id: String,
    mount_point: String,
    export_service: State<'_, Arc<ExportService>>,
    checkpoint_service: State<'_, Arc<CheckpointService>>,
    app_handle: AppHandle,
) -> Result<ExportResult> {
    // Get the pending checkpoint
    let checkpoint = checkpoint_service
        .get_pending_checkpoint(&device_id)?
        .ok_or_else(|| crate::error::CrateError::Export("No pending checkpoint found".to_string()))?;

    // Create a request from the checkpoint
    let request = ExportRequest {
        device_id: checkpoint.device_id.clone(),
        mount_point,
        device_name: checkpoint.device_name.clone(),
        playlist_ids: checkpoint.playlist_ids.clone(),
        enable_sync: true,
        use_device_library_plus: false,
    };

    // Resume the export - the export service will detect the checkpoint
    // and skip already-completed tracks
    let result = export_service.export_playlists(&app_handle, request)?;

    // If successful, delete the checkpoint
    if result.success {
        checkpoint_service.complete_checkpoint(&checkpoint.id)?;
    }

    Ok(result)
}
